notification/app/notification/controllers/Main.scala

```diff
@@ -72,5 +62,28 @@ final class Main @Inject()(
     val push = Push(request.body, Right(UserId(userId)))
     pushGeneric(push)
   }
+  
+  private def pushGeneric(push: Push) = {
+    notificationSender.sendNotification(push) flatMap {
+      case \/-(report) =>
+        logger.info(s"Notification was sent: $push")
+        notifyAboutPush(trackingObservers, report) map {
+          case \/-(_) => Created(Json.toJson(PushResult.fromReport(report)))
+          case -\/(errors) => Created(Json.toJson(PushResult.fromReport(report).withTrackingErrors(errors.map(_.reason))))
+        }
+      case -\/(error) =>
+        logger.error(s"Notification ($push) could not be sent: $error")
+        Future.successful(handleErrors(error))
+    }
+  }
 
+  private def notifyAboutPush(observers: Seq[TrackingObserver], report: NotificationReport) = {
+    def notifyObservers = observers.map { _.notificationSent(report) }
+    val noErrors = ().right[List[TrackingError]]
```

Comment:
This might be clear as:
```
val NoErrors = ().right[List[TrackingError]]
Future.fold(notifyObservers)(NoErrors) {
  case (NoErrors, -\/(err)) => List(err).left
  case (-\/(errors), -\/(err)) => (err :: errors).left
  case (aggr, _) => aggr
}
```

or even more radically using something like:
```  
private def aggregateFailures(ls: Seq[TrackerResult[Unit]]): Seq[TrackingError] \/ Unit = ls collect {
  case -\/(err) => err
} match {
  case errors if errors.nonEmpty => errors.left
  case Nil => ().right
}
```